### PR TITLE
Cross Validation and TrainTest

### LINKED ISSUE #6: Enable cross-validation for LearningPipeline
Cross-validation would make it easier to validate models if the data is not already split into train and test files (and generally helps give more precise metrics). This is already exposed as an entrypoint [here](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML/Runtime/EntryPoints/CrossValidationMacro.cs), but needs to be enabled for LearningPipeline.

### PR DESCRIPTION
**Pull request by codemzs**
Introduces Cross Validation and Train-Test evaluators. and fixes #6 #251

### CODE REVIEW

File: src/Microsoft.ML/ILearningPipelineItem.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@ namespace Microsoft.ML
     public interface ILearningPipelineItem
     {
         ILearningPipelineStep ApplyStep(ILearningPipelineStep previousStep, Experiment experiment);
+        Var<IDataView> GetInputData();
```
Comment (glebuk) [inline src/Microsoft.ML/ILearningPipelineItem.cs]: >GetInputData [](start = 23, length = 12)

Please add xml comment. Not 100% clear what this fucntion supposed to do:  input to the item? When will this be useful?

**Comment by glebuk** [comment]
>         void SetInput(IHostEnvironment environment, Experiment experiment);

What does GetInputData() mean for a loader?  Text loader has no input IDV to it...

---
Refers to: src/Microsoft.ML/ILearningPipelineItem.cs:25 in 0ef9ce1. [](commit_id = 0ef9ce16f0e250eac6ffc6470e5722cf1b6fd202, deletion_comment = False)

File: src/Microsoft.ML/Models/BinaryClassificationEvaluator.cs [hunk 2]
```diff
@@ -23,7 +24,7 @@ public sealed partial class BinaryClassificationEvaluator
         /// <returns>
         /// A BinaryClassificationMetrics instance that describes how well the model performed against the test data.
         /// </returns>
-        public BinaryClassificationMetrics Evaluate(PredictionModel model, ILearningPipelineLoader testData)
+        public List<BinaryClassificationMetrics> Evaluate(PredictionModel model, ILearningPipelineLoader testData)
```
Comment (glebuk) [inline src/Microsoft.ML/Models/BinaryClassificationEvaluator.cs]: >Evaluate [](start = 49, length = 8)

Why would a singe model produce a list of metrics? What does the list index corresponds to?

Comment (glebuk) [reply src/Microsoft.ML/Models/BinaryClassificationEvaluator.cs]: The list makes sense if you have a list of input models, but not if there is only one.

---
In reply to: [190402103](https://github.com/dotnet/machinelearning/pull/212#discussion_r190402103) [](ancestors = 190402103)

**Comment by glebuk** [comment]
>     public sealed class BinaryClassificationMetrics

It seems very strange that a single class that represents a matrix now has to also be a  collection.
Have you considered instead, to keep those classes as is, and instead just give the IRow as input, not IDV.

---
Refers to: src/Microsoft.ML/Models/BinaryClassificationMetrics.cs:16 in 0ef9ce1. [](commit_id = 0ef9ce16f0e250eac6ffc6470e5722cf1b6fd202, deletion_comment = False)

File: src/Microsoft.ML/Models/CrossValidator.cs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.ML.Models
+{
+    public sealed partial class CrossValidator
+    {
+        public CrossValidationOutput<TInput, TOutput> CrossValidate<TInput, TOutput>(LearningPipeline pipeline)
```
Comment (glebuk) [inline src/Microsoft.ML/Models/CrossValidator.cs]: >CrossValidate [](start = 54, length = 13)

Do we have a version of this for IDV as input?

**Comment by glebuk** [comment]
>                 public Var<Microsoft.ML.Runtime.Data.IDataView> ConfusionMatrix { get; set; } = new Var<Microsoft.ML.Runtime.Data.IDataView>();

Better to keep it as it was - return arrayvar. See comment on binaryclassificationmetric

---
Refers to: src/Microsoft.ML/CSharpApi.cs:2285 in 0ef9ce1. [](commit_id = 0ef9ce16f0e250eac6ffc6470e5722cf1b6fd202, deletion_comment = False)

**Comment by codemzs** [comment]
>         void SetInput(IHostEnvironment environment, Experiment experiment);

yeah so it will return null...

---
In reply to: [391499141](https://github.com/dotnet/machinelearning/pull/212#issuecomment-391499141) [](ancestors = 391499141)

---
Refers to: src/Microsoft.ML/ILearningPipelineItem.cs:25 in 0ef9ce1. [](commit_id = 0ef9ce16f0e250eac6ffc6470e5722cf1b6fd202, deletion_comment = False)

**Comment by codemzs** [comment]
>                 public Var<Microsoft.ML.Runtime.Data.IDataView> ConfusionMatrix { get; set; } = new Var<Microsoft.ML.Runtime.Data.IDataView>();

This is from Yael's change which you approved. 

---
In reply to: [391502108](https://github.com/dotnet/machinelearning/pull/212#issuecomment-391502108) [](ancestors = 391502108)

---
Refers to: src/Microsoft.ML/CSharpApi.cs:2285 in 0ef9ce1. [](commit_id = 0ef9ce16f0e250eac6ffc6470e5722cf1b6fd202, deletion_comment = False)

File: src/Microsoft.ML/Models/BinaryClassificationMetrics.cs [hunk 4]
```diff
@@ -18,7 +19,7 @@ private BinaryClassificationMetrics()
         {
         }
 
-        internal static BinaryClassificationMetrics FromMetrics(IHostEnvironment env, IDataView overallMetrics, IDataView confusionMatrix)
+        internal static List<BinaryClassificationMetrics> FromMetrics(IHostEnvironment env, IDataView overallMetrics, IDataView confusionMatrix)
```
[the remark this answers is not in the document]
Comment (codemzs) [reply src/Microsoft.ML/Models/BinaryClassificationMetrics.cs]: This is not related to my change. "Overallmetrics" is consistent with what Idataview object is named in the runtime. 

---
In reply to: [190403397](https://github.com/dotnet/machinelearning/pull/212#discussion_r190403397) [](ancestors = 190403397)

**Comment by codemzs** [comment]
>     public sealed class BinaryClassificationMetrics

Why is it strange? Can you please clarify? 

---
In reply to: [391500787](https://github.com/dotnet/machinelearning/pull/212#issuecomment-391500787) [](ancestors = 391500787)

---
Refers to: src/Microsoft.ML/Models/BinaryClassificationMetrics.cs:16 in 0ef9ce1. [](commit_id = 0ef9ce16f0e250eac6ffc6470e5722cf1b6fd202, deletion_comment = False)

File: src/Microsoft.ML/Models/BinaryClassificationEvaluator.cs (hunk 2 above)
Comment (codemzs) [reply src/Microsoft.ML/Models/BinaryClassificationEvaluator.cs]: Correct. My bad.

---
In reply to: [190402722](https://github.com/dotnet/machinelearning/pull/212#discussion_r190402722) [](ancestors = 190402722,190402103)

File: src/Microsoft.ML/Models/CrossValidator.cs (hunk 3 above)
Comment (codemzs) [reply src/Microsoft.ML/Models/CrossValidator.cs]: If by IDV you mean DataCollection, then yes the user can insert a data collection object in the pipeline and do CV on it.

---
In reply to: [190404310](https://github.com/dotnet/machinelearning/pull/212#discussion_r190404310) [](ancestors = 190404310)

File: src/Microsoft.ML/Data/TextLoader.cs [hunk 5]
```diff
@@ -21,7 +21,6 @@ public TextLoaderRange()
         /// <summary>
         /// Convenience constructor for the scalar case, when a given column 
         /// in the schema spans only a single column in the dataset.
-        /// <see cref="Min"/> and <see cref="Max"/> are set to the single value <paramref name="ordinal"/>.
```
Comment (codemzs) [inline src/Microsoft.ML/Data/TextLoader.cs]: >/// <see cref="Min"/> and <see cref="Max"/> are set to the single value <paramref name="ordinal"/>. [](start = 8, length = 99)

Not sure how this got removed.

File: test/Microsoft.ML.Tests/Scenarios/SentimentPredictionTests.cs [hunk 6]
```diff
@@ -23,6 +23,9 @@ public partial class ScenariosTests
         [Fact]
         public void TrainAndPredictSentimentModelTest()
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/Scenarios/SentimentPredictionTests.cs]: >TrainAndPredictSentimentModelTest [](start = 20, length = 33)

I would advise to split this 250 lines test into subtests.

File: src/Microsoft.ML/ILearningPipelineItem.cs (hunk 1 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML/ILearningPipelineItem.cs]: >When will this be useful
Is still remain open question

---
In reply to: [190401291](https://github.com/dotnet/machinelearning/pull/212#discussion_r190401291) [](ancestors = 190401291)

File: src/Microsoft.ML/Data/CollectionDataSource.cs [hunk 7]
```diff
@@ -52,6 +52,8 @@ public void SetInput(IHostEnvironment environment, Experiment experiment)
                 experiment.SetInput(_dataViewEntryPoint.Data, _dataView);
             }
 
+            public Var<IDataView> GetInputData() => null;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML/Data/CollectionDataSource.cs]: >public Var<IDataView> GetInputData() => null; [](start = 12, length = 45)

why I don't see this change in TextLoader, it also suppose to implement ILearningPipeline?

File: ZBaselines/Common/EntryPoints/core_ep-list.tsv [hunk 8]
```diff
@@ -3,6 +3,7 @@ Data.DataViewReference	Pass dataview from memory to experiment	Microsoft.ML.Runt
 Data.IDataViewArrayConverter	Create and array variable	Microsoft.ML.Runtime.EntryPoints.CrossValidationBinaryMacro	MakeArray	Microsoft.ML.Runtime.EntryPoints.CrossValidationBinaryMacro+ArrayIDataViewInput	Microsoft.ML.Runtime.EntryPoints.CrossValidationBinaryMacro+ArrayIDataViewOutput
 Data.PredictorModelArrayConverter	Create and array variable	Microsoft.ML.Runtime.EntryPoints.CrossValidationBinaryMacro	MakeArray	Microsoft.ML.Runtime.EntryPoints.CrossValidationBinaryMacro+ArrayIPredictorModelInput	Microsoft.ML.Runtime.EntryPoints.CrossValidationBinaryMacro+ArrayIPredictorModelOutput
 Data.TextLoader	Import a dataset from a text file	Microsoft.ML.Runtime.EntryPoints.ImportTextData	TextLoader	Microsoft.ML.Runtime.EntryPoints.ImportTextData+LoaderInput	Microsoft.ML.Runtime.EntryPoints.ImportTextData+Output
+Data.TransformModelArrayConverter	Create and array variable	Microsoft.ML.Runtime.EntryPoints.CrossValidationBinaryMacro	MakeArray	Microsoft.ML.Runtime.EntryPoints.CrossValidationBinaryMacro+ArrayITransformModelInput	Microsoft.ML.Runtime.EntryPoints.CrossValidationBinaryMacro+ArrayITransformModelOutput
```
Comment (Ivanidzo4ka) [inline ZBaselines/Common/EntryPoints/core_ep-list.tsv]: >Create and array variable [](start = 34, length = 25)

So what's the difference between this one and Data.PredictorModelArrayConverter and Data.IDataViewArrayConverter. All of them "Create and array variable"

File: src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs [hunk 9]
```diff
@@ -23,8 +24,14 @@ public sealed class SubGraphInput
...
         {
-            [Argument(ArgumentType.Required, HelpText = "The model", SortOrder = 1)]
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The model", SortOrder = 1)]
             public Var<IPredictorModel> Model;
+            
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Transform model", SortOrder = 2)]
+            public Var<ITransformModel> TransformModel;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Indicates to use transform model instead of predictor model.", SortOrder = 3)]
+            public bool UseTransformModel = false;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs]: Unrelated to your code, but can anyone explain what's the difference between IPredictorModel and ITransformModel? or maybe related, because HelpText not helping at all. "The model" vs "Transform Model". What should I choose? hmmm....

File: src/Microsoft.ML/Models/TrainTestEvaluator.cs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                TestingData = (testData.ApplyStep(null, experiment) as ILearningPipelineDataStep).Data;
+                Nodes = subGraph;
+                TransformModel = null;
+                Inputs.Data = firstTransform.GetInputData();
+                Outputs.TransformModel = lastTransformModel;
+                Outputs.UseTransformModel = true;
+                var crossValidateOutput = experiment.Add(this);
+                experiment.Compile();
+                foreach (ILearningPipelineLoader loader in loaders)
+                {
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML/Models/TrainTestEvaluator.cs]: >{ [](start = 16, length = 1)

We used to hate braces for single line. Maybe times change.

File: src/Microsoft.ML/Models/TrainTestEvaluator.cs [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+                    loader.SetInput(environment, experiment);
+                }
+                testData.SetInput(environment, experiment);
+
+                experiment.Run();
+
+                TrainTestEvaluatorOutput<TInput, TOutput> tteo = new TrainTestEvaluatorOutput<TInput, TOutput>();
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML/Models/TrainTestEvaluator.cs]: >TrainTestEvaluatorOutput<TInput, TOutput> tteo  [](start = 16, length = 47)

var result or var output or var evalOutput.

**Comment by Ivanidzo4ka** [comment]
>                 subGraphNodes.Add(EntryPointNode.Create(env, "Models.TrainTestEvaluator", args, node.Catalog, node.Context, inputBindingMap, inputMap, outputMap));

Can I suggest you to make it string const?

---
Refers to: src/Microsoft.ML/Runtime/EntryPoints/CrossValidationMacro.cs:297 in 3991971. [](commit_id = 3991971130a0c88099becca33cb7ba1b0cd29e84, deletion_comment = False)

File: src/Microsoft.ML/Runtime/EntryPoints/CrossValidationMacro.cs [hunk 12]
```diff
@@ -403,6 +476,7 @@ private static IMamlEvaluator GetEvaluator(IHostEnvironment env, MacroUtils.Trai
                     return new MultiOutputRegressionMamlEvaluator(env, new MultiOutputRegressionMamlEvaluator.Arguments());
                 default:
                     throw env.ExceptParam(nameof(kind), $"Trainer kind {kind} does not have an evaluator");
+
             }
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML/Runtime/EntryPoints/CrossValidationMacro.cs]: why?

File: src/Microsoft.ML/Data/CollectionDataSource.cs (hunk 7 above)
Comment (codemzs) [reply src/Microsoft.ML/Data/CollectionDataSource.cs]: Its code generated there in CSharpAPI.cs

---
In reply to: [191002526](https://github.com/dotnet/machinelearning/pull/212#discussion_r191002526) [](ancestors = 191002526)

File: src/Microsoft.ML/Models/TrainTestEvaluator.cs (hunk 10 above)
Comment (codemzs) [reply src/Microsoft.ML/Models/TrainTestEvaluator.cs]: lol

---
In reply to: [191004971](https://github.com/dotnet/machinelearning/pull/212#discussion_r191004971) [](ancestors = 191004971)

File: ZBaselines/Common/EntryPoints/core_ep-list.tsv (hunk 8 above)
Comment (codemzs) [reply ZBaselines/Common/EntryPoints/core_ep-list.tsv]: they create array variables of different types, like ITranformModel, IPredictorModel, IDV etc. 

---
In reply to: [191003203](https://github.com/dotnet/machinelearning/pull/212#discussion_r191003203) [](ancestors = 191003203)

File: src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs (hunk 9 above)
Comment (codemzs) [reply src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs]: I'll change The Model to The predictor model.

---
In reply to: [191004407](https://github.com/dotnet/machinelearning/pull/212#discussion_r191004407) [](ancestors = 191004407)

File: src/Microsoft.ML/Models/BinaryClassificationEvaluator.cs [hunk 13]
```diff
@@ -66,7 +68,11 @@ public BinaryClassificationMetrics Evaluate(PredictionModel model, ILearningPipe
                     throw environment.Except($"Could not find ConfusionMatrix in the results returned in {nameof(BinaryClassificationEvaluator)} Evaluate.");
                 }
 
-                return BinaryClassificationMetrics.FromMetrics(environment, overallMetrics, confusionMatrix);
+                var metric = BinaryClassificationMetrics.FromMetrics(environment, overallMetrics, confusionMatrix);
+
+                Contracts.Assert(metric.Count == 1);
+
+                return metric.First();
```
[the remark this answers is not in the document]
Comment (codemzs) [reply src/Microsoft.ML/Models/BinaryClassificationEvaluator.cs]: It returns a list because that function is retrieves rows from an idataview that contains metrics, for evaluation of a single model it will contain one row but for cross validation it will contain multiple rows that represent all the folds. This function is for single model evaluation case so I take out the first row and only expect one row. Does that answer your question?

---
In reply to: [191001670](https://github.com/dotnet/machinelearning/pull/212#discussion_r191001670) [](ancestors = 191001670)

**Comment by codemzs** [comment]
>                 subGraphNodes.Add(EntryPointNode.Create(env, "Models.TrainTestEvaluator", args, node.Catalog, node.Context, inputBindingMap, inputMap, outputMap));

This is a function call, how can I make it a const?

---
In reply to: [392181249](https://github.com/dotnet/machinelearning/pull/212#issuecomment-392181249) [](ancestors = 392181249)

---
Refers to: src/Microsoft.ML/Runtime/EntryPoints/CrossValidationMacro.cs:297 in 3991971. [](commit_id = 3991971130a0c88099becca33cb7ba1b0cd29e84, deletion_comment = False)

File: src/Microsoft.ML/Models/BinaryClassificationEvaluator.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Ivanidzo4ka) [reply src/Microsoft.ML/Models/BinaryClassificationEvaluator.cs]: ok, let me rephrase my question, why you do metric.First() instead of metric[0]? 
btw metric.First() will throw if metric is empty which you wouldn't catch through Contracts.Assert in Release build

---
In reply to: [191019339](https://github.com/dotnet/machinelearning/pull/212#discussion_r191019339) [](ancestors = 191019339,191001670)

**Comment by Ivanidzo4ka** [comment]
>                 subGraphNodes.Add(EntryPointNode.Create(env, "Models.TrainTestEvaluator", args, node.Catalog, node.Context, inputBindingMap, inputMap, outputMap));

I meant "Models.TrainTestEvaluator" string. if you someone change that string inside entrypoint definition, they also have to change this string which is error prone

---
In reply to: [392203208](https://github.com/dotnet/machinelearning/pull/212#issuecomment-392203208) [](ancestors = 392203208,392181249)

---
Refers to: src/Microsoft.ML/Runtime/EntryPoints/CrossValidationMacro.cs:297 in 3991971. [](commit_id = 3991971130a0c88099becca33cb7ba1b0cd29e84, deletion_comment = False)

File: src/Microsoft.ML/Models/TrainTestEvaluator.cs (hunk 11 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML/Models/TrainTestEvaluator.cs]: i would still prefer var.

---
In reply to: [191005451](https://github.com/dotnet/machinelearning/pull/212#discussion_r191005451) [](ancestors = 191005451)

File: src/Microsoft.ML/Models/CrossValidator.cs [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    /// </summary>
+    public sealed partial class CrossValidator
+    {
+        /// <summary>
+        /// Performs cross validation on a pipeline.
+        /// </summary>
+        /// <typeparam name="TInput">Class type that represents input schema.</typeparam>
+        /// <typeparam name="TOutput">Class type that represents prediction schema.</typeparam>
+        /// <param name="pipeline">Machine learning pipeline that contain may contain loader, transforms and at least one trainer.</param>
+        /// <returns>List containning metrics and predictor model for each fold</returns>
```
Comment (zeahmed) [inline src/Microsoft.ML/Models/CrossValidator.cs]: containning -> containing

File: src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs [hunk 15]
```diff
@@ -23,8 +24,14 @@ public sealed class SubGraphInput
 
         public sealed class SubGraphOutput
         {
-            [Argument(ArgumentType.Required, HelpText = "The model", SortOrder = 1)]
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The predictor model", SortOrder = 1)]
             public Var<IPredictorModel> Model;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs]: >Model [](start = 40, length = 5)

Can you name it PredictorModel?
it's confusing if you start read code
in one place you have this:
            varName = input.Outputs.UseTransformModel ? input.Outputs.TransformModel.VarName : input.Outputs.Model.VarName;
in other place
string outputVarName = input.Outputs.UseTransformModel ? node.GetOutputVariableName(nameof(Output.TransformModel)) : 
                node.GetOutputVariableName(nameof(Output.PredictorModel));

File: src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs [hunk 16]
```diff
@@ -117,10 +129,13 @@ public sealed class Output
             subGraphRunContext.RemoveVariable(dataVariable);
 
             // Change the subgraph to use the model variable as output.
-            varName = input.Outputs.Model.VarName;
+            varName = input.Outputs.UseTransformModel ? input.Outputs.TransformModel.VarName : input.Outputs.Model.VarName;
             if (!subGraphRunContext.TryGetVariable(varName, out dataVariable))
                 throw env.Except($"Invalid variable name '{varName}'.");
-            string outputVarName = node.GetOutputVariableName(nameof(Output.PredictorModel));
+
+            string outputVarName = input.Outputs.UseTransformModel ? node.GetOutputVariableName(nameof(Output.TransformModel)) :
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs]: >node.GetOutputVariableName [](start = 69, length = 26)

you can probably move it forward in your statement and put ternary operator inside it.
node.GetOuputVariableName( input.Outputs.UseTransformModel? nameof(Output.TransformModel): nameof(Output.PredictorModel));

File: src/Microsoft.ML/Models/TrainTestEvaluator.cs [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+    /// </summary>
+    public sealed partial class TrainTestEvaluator
+    {
+        /// <summary>
+        /// Performs train-test on a pipeline.
+        /// </summary>
+        /// <typeparam name="TInput">Class type that represents input schema.</typeparam>
+        /// <typeparam name="TOutput">Class type that represents prediction schema.</typeparam>
+        /// <param name="pipeline">Machine learning pipeline that contains <see cref="ILearningPipelineLoader"/>, 
+        /// transforms and at least one trainer.</param>
```
Comment (zeahmed) [inline src/Microsoft.ML/Models/TrainTestEvaluator.cs]: `at most one trainer`??

File: test/Microsoft.ML.Tests/Scenarios/SentimentPredictionTests.cs [hunk 18]
```diff
@@ -143,6 +143,273 @@ public void TrainAndPredictSentimentModelTest()
...
+            {
+                new SentimentData
+                {
+                    SentimentText = "Please refrain from adding nonsense to Wikipedia."
+                },
+                new SentimentData
+                {
+                    SentimentText = "He is a CHEATER, and the article should say that."
+                }
+            };
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/Scenarios/SentimentPredictionTests.cs]: is this a same code in all 3 tests? Can you refactor it?

Comment (codemzs) [reply test/Microsoft.ML.Tests/Scenarios/SentimentPredictionTests.cs]: Its ok for tests. 

---
In reply to: [191030443](https://github.com/dotnet/machinelearning/pull/212#discussion_r191030443) [](ancestors = 191030443)

File: src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs (hunk 16 above)
Comment (codemzs) [reply src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs]: this is just fine.

---
In reply to: [191029906](https://github.com/dotnet/machinelearning/pull/212#discussion_r191029906) [](ancestors = 191029906)

File: src/Microsoft.ML/Models/CrossValidator.cs [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+    /// Performs cross-validation on a pipeline.
+    /// </summary>
+    public sealed partial class CrossValidator
+    {
+        /// <summary>
+        /// Performs cross validation on a pipeline.
+        /// </summary>
+        /// <typeparam name="TInput">Class type that represents input schema.</typeparam>
+        /// <typeparam name="TOutput">Class type that represents prediction schema.</typeparam>
+        /// <param name="pipeline">Machine learning pipeline that contain may contain loader, transforms and at least one trainer.</param>
```
[the remark this answers is not in the document]
Comment (codemzs) [reply src/Microsoft.ML/Models/CrossValidator.cs]: You can do stacking. Pipeline framework supports that.

---
In reply to: [191029184](https://github.com/dotnet/machinelearning/pull/212#discussion_r191029184) [](ancestors = 191029184)

File: ZBaselines/Common/EntryPoints/core_manifest.json [hunk 20]
```diff
@@ -1476,6 +1524,14 @@
           },
           "Desc": "The final model including the trained predictor model and the model from the transforms, provided as the Input.TransformModel."
         },
+        {
+          "Name": "TransformModel",
+          "Type": {
+            "Kind": "Array",
+            "ItemType": "TransformModel"
+          },
+          "Desc": "The final model including the trained predictor model and the model from the transforms, provided as the Input.TransformModel."
```
[the remark this answers is not in the document]
Comment (codemzs) [reply ZBaselines/Common/EntryPoints/core_manifest.json]: Yeah, so we introduced a second option in CV and TrainTest where the final model is of ITransformModel.

---
In reply to: [191033404](https://github.com/dotnet/machinelearning/pull/212#discussion_r191033404) [](ancestors = 191033404)

File: src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs [hunk 21]
```diff
@@ -62,31 +69,36 @@ public sealed class Arguments
...
         public sealed class Output
         {
-            [TlcModule.Output(Desc = "The final model including the trained predictor model and the model from the transforms, provided as the Input.TransformModel.", SortOrder = 1)]
+            [TlcModule.Output(Desc = "The final model including the trained predictor model and the model from the transforms, " +
+                "provided as the Input.TransformModel.", SortOrder = 1)]
             public IPredictorModel PredictorModel;
 
-            [TlcModule.Output(Desc = "Warning dataset", SortOrder = 2)]
+            [TlcModule.Output(Desc = "The final model including the trained predictor model and the model from the transforms, " +
+                "provided as the Input.TransformModel.", SortOrder = 2)]
```
Comment (sfilipi) [inline src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs]: Update description.

File: src/Microsoft.ML/Models/BinaryClassificationMetrics.cs [hunk 22]
```diff
@@ -18,41 +19,53 @@ private BinaryClassificationMetrics()
...
             env.AssertValue(overallMetrics);
             env.AssertValue(confusionMatrix);
 
             var metricsEnumerable = overallMetrics.AsEnumerable<SerializationClass>(env, true, ignoreMissingColumns: true);
             var enumerator = metricsEnumerable.GetEnumerator();
+
             if (!enumerator.MoveNext())
             {
-                throw env.Except("The overall RegressionMetrics didn't have any rows.");
+                throw env.Except("The overall RegressionMetrics didn't have sufficient rows.");
```
Comment (TomFinley) [inline src/Microsoft.ML/Models/BinaryClassificationMetrics.cs]: >sufficient [](start = 76, length = 10)

The old error message was more specific and accurate. You don't mention what *would* be sufficient. "Any" makes this clear.

File: src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs [hunk 23]
```diff
@@ -23,8 +24,14 @@ public sealed class SubGraphInput
...
         {
-            [Argument(ArgumentType.Required, HelpText = "The model", SortOrder = 1)]
-            public Var<IPredictorModel> Model;
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The predictor model", SortOrder = 1)]
+            public Var<IPredictorModel> PredictorModel;
+            
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Transform model", SortOrder = 2)]
+            public Var<ITransformModel> TransformModel;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Indicates to use transform model instead of predictor model.", SortOrder = 3)]
```
Comment (sfilipi) [inline src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs]: Why is this necessary? There is a convention for the graph variables: if set to null/absent don't use it. 
Is there any case where the TransformModel is provided but it shouldn't get used?

If so, please add a comment to code.

File: ZBaselines/Common/EntryPoints/core_manifest.json (hunk 20 above)
[the remark this answers is not in the document]
Comment (sfilipi) [reply ZBaselines/Common/EntryPoints/core_manifest.json]: Update the part of the comment that talks about bundling the predictor model in it.

[the remark this answers is not in the document]
Comment (codemzs) [reply ZBaselines/Common/EntryPoints/core_manifest.json]: Can you please clarify? The comment already talks about predictor being included in the final model.

---
In reply to: [191049219](https://github.com/dotnet/machinelearning/pull/212#discussion_r191049219) [](ancestors = 191049219)

File: src/Microsoft.ML/Models/CrossValidator.cs [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+                            environment,
+                            experiment.GetOutput(crossValidateOutput.OverallMetrics),
+                            experiment.GetOutput(crossValidateOutput.ConfusionMatrix), 2);
+                    }
+                    else if(Kind == MacroUtilsTrainerKinds.SignatureMultiClassClassifierTrainer)
+                    {
+                        cvOutput.ClassificationMetrics = ClassificationMetrics.FromMetrics(
+                            environment,
+                            experiment.GetOutput(crossValidateOutput.OverallMetrics),
+                            experiment.GetOutput(crossValidateOutput.ConfusionMatrix), 2);
```
[the remark this answers is not in the document]
Comment (codemzs) [reply src/Microsoft.ML/Models/CrossValidator.cs]: First two rows are average and std. deviation of the folds and they don't contain any confusion matrix.

---
In reply to: [191048669](https://github.com/dotnet/machinelearning/pull/212#discussion_r191048669) [](ancestors = 191048669)

File: src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs (hunk 23 above)
Comment (codemzs) [reply src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs]: The code generated entrypoint class always initializes the variables(see csharpapi.cs in this PR), so a user may set TransformModel but PredictorModel would also been initialized to a non null value. Hence I cannot know which one the user had set. As of today the user has to explicitly set variables to null that they don't want the code to use for example in TrainTestMacro arguments TransformModel needs to be set to null explicitly each time a graph is created even though in the code it is initialized to null. 

You could argue why not adopt the same model of making the user set values to null that they don't want the code to use and that might be a fair argument as it would clean up the code a little.


---
In reply to: [191048298](https://github.com/dotnet/machinelearning/pull/212#discussion_r191048298) [](ancestors = 191048298)

**Comment by codemzs** [comment]
@dotnet-bot rebuild
